lockbox-ios/Presenter/RootPresenter.swift

```diff
@@ -62,39 +75,32 @@ class RootPresenter {
         self.telemetryActionHandler = telemetryActionHandler
         self.biometryManager = biometryManager
 
+        // todo: update tests with populated oauth and profile info
         Observable.combineLatest(self.accountStore.oauthInfo, self.accountStore.profile)
-            .bind { latest in
-                    if let oauthInfo = latest.0,
-                        let profile = latest.1 {
+                .map { OAuthProfile(oauthInfo: $0.0, profile: $0.1) }
+                .distinctUntilChanged()
+                .bind { latest in
+                    if let oauthInfo = latest.oauthInfo,
+                        let profile = latest.profile {
                         self.dataStoreActionHandler.invoke(.updateCredentials(oauthInfo: oauthInfo, fxaProfile: profile))
+                    } else if latest.oauthInfo == nil && latest.profile == nil {
+                        self.routeActionHandler.invoke(LoginRouteAction.welcome)
+                        self.dataStoreActionHandler.invoke(.reset)
                     }
                 }
                 .disposed(by: self.disposeBag)
 
-        self.dataStore.storageState
-            .subscribe(onNext: { storageState in
-                    switch storageState {
-                    case .Unprepared, .Locked:
-                        self.routeActionHandler.invoke(LoginRouteAction.welcome)
-                    case .Unlocked:
-                        self.routeActionHandler.invoke(MainRouteAction.list)
-                    default:
-                        break
-                    }
+        Observable.combineLatest(self.dataStore.locked, self.accountStore.hasOldAccountInformation, self.dataStore.storageState)
+                .filter { !$0.1 && $0.2 != LoginStoreState.Unprepared }
+                .map { $0.0 }
+                .distinctUntilChanged()
+                .subscribe(onNext: { locked in
+                    let route: RouteAction = locked ? LoginRouteAction.welcome : MainRouteAction.list
+
+                    self.routeActionHandler.invoke(route)
```

Comment:
This looks… brittle.

I'm not sure I understand why we don't separate out the `hasOldAccountInformation` to a `take(1)`.